### PR TITLE
crimson/osd: add "ceph tell <pgid> <command>" support

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
at this moment, we support both

```
ceph pg 1.0 query
```

and 
```
ceph tell 1.0 query
```

but since, in `OSD.cc`,

>  // old form: ceph pg <pgid> command ...

and

>  // new form: tell <pgid> <cmd> for both cli and rest

i assume that we will deprecate the former in long term. if we have to support the former one when releasing crimson, we probably could have a shim in ceph-mgr to translate the latter to the former and forward the command to osd.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/admin/pg_commands.cc [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  virtual seastar::future<tell_result_t>
+  do_command(Ref<PG> pg,
+             const cmdmap_t& cmdmap,
+             std::string_view format,
+             ceph::bufferlist&& input) const = 0;
+
+  OSD& osd;
+};
+
+class QueryCommand final : public PGCommand {
```
Comment (tchaikov) [inline src/crimson/admin/pg_commands.cc:81]: `QueryCommand` is not a `PGCommand`, but it is a `AdminSocketHook`. i could use CRTP to make this fact more obvious. but that'd be overkill.

File: src/crimson/admin/pg_commands.cc [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    std::string pgid_str;
+    pg_t pgid;
+    if (!cmd_getval(cmdmap, "pgid", pgid_str)) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, "no pgid specified"});
+    } else if (!pgid.parse(pgid_str.c_str())) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, fmt::format("couldn't parse pgid '{}'", pgid_str)});
+    }
+    // am i the primary for this pg?
```
Comment (ronen-fr) [inline src/crimson/admin/pg_commands.cc:52]: I would suggest that if we encounter this check ("am I the primary") a second time, this block
of checks should be factored out. Not urgent.

**Review by ronen-fr — APPROVED** [review]

File: src/crimson/admin/pg_commands.cc (hunk 1 above)
Comment (ronen-fr) [reply src/crimson/admin/pg_commands.cc:81]: Adding a comment?

Comment (tchaikov) [reply src/crimson/admin/pg_commands.cc:81]: i'd prefer not adding a comment. i think it's more confusing if i failed to send the right message.

File: src/crimson/admin/pg_commands.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/admin/pg_commands.cc:52]: could you be more specific? what change do you suggest? `PGCommand` tries to extract the shared code among pg tell commands. what kind of "second time" we are expecting?

File: src/crimson/admin/pg_commands.cc [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+              " removing or reverting to a prior version if one is"
+              " available"}
+  {}
+  seastar::future<tell_result_t>
+  do_command(Ref<PG> pg,
+             const cmdmap_t& cmdmap,
+             std::string_view,
+             ceph::bufferlist&&) const final
+  {
+    // what do to with the unfound object specifically.
```
Comment (tchaikov) [inline src/crimson/admin/pg_commands.cc:124]:
```suggestion
    // what to do with the unfound object specifically.
```

**Comment by tchaikov** [comment]
changelog

- fix the syntax error in comment.

File: src/crimson/admin/pg_commands.cc (hunk 2 above)
Comment (ronen-fr) [reply src/crimson/admin/pg_commands.cc:52]: Sure. I was ref the code that starts with a pgid, and:
- fetches the osd map,
- verifies that the named pg exists (according to the map), and
- that we are part of the PG's set, and
- we are the primary,

returning either a ref<pg> or an error.

I suspect that this is a common test, and we may want to have it as separate method (not necessarily returning a tell_t, as that indeed would limit applicability).

Comment (tchaikov) [reply src/crimson/admin/pg_commands.cc:52]: Thanks Ronen, I understand you concern now. But I am not sure how to address it with a balanced way between “don’t repeat yourself” and too much indirection layers. As this is two lines of code..

Could you sent a follow-up PR to help me get this right? Much appreciated.

Comment (ronen-fr) [reply src/crimson/admin/pg_commands.cc:52]: WIll do - the minute I encounter a second usage of that...
More seriously - will indeed do that once I return to Crimson code.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-09-06_13:32:55-rados-wip-kefu-testing-2020-09-06-2017-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-07_01:57:21-rados-wip-kefu-testing-2020-09-06-2017-distro-basic-smithi

failures tracked by 

- https://tracker.ceph.com/issues/47306
- https://tracker.ceph.com/issues/47179

**Pull request merged**